### PR TITLE
Fix #73 - Calcs now correctly the TFLOPS for MacBook Pro M1 Max

### LINKED ISSUE #73: Showing 0 TFLOPS for MacBook Pro node
image

### PR DESCRIPTION
**Pull request by stephanj**

### CODE REVIEW

**Comment by AlexCheema** [comment]
Awesome work! Thanks so much for the PR.

Could you also update the tests in `exo/topology/test_device_capabilities.py` @stephanj

File: exo/topology/device_capabilities.py [hunk 1]
```diff
@@ -77,24 +76,60 @@ def device_capabilities() -> DeviceCapabilities:
...
-    memory_units = memory_str.split()
-    memory_value = int(memory_units[0])
-    if memory_units[1] == "GB":
-        memory = memory_value * 1024
-    else:
-        memory = memory_value
-
-    # Assuming static values for other attributes for demonstration
-    return DeviceCapabilities(model=model_id, chip=chip_id, memory=memory, flops=CHIP_FLOPS.get(chip_id, DeviceFlops(fp32=0, fp16=0, int8=0)))
+    hw_info = subprocess.check_output(['system_profiler', 'SPHardwareDataType']).decode('utf-8')
```
Comment (AlexCheema) [inline exo/topology/device_capabilities.py:86]: I just took another look and there's a `-json` flag that might make parsing a bit cleaner: `system_profiler SPHardwareDataType -json`

File: exo/topology/device_capabilities.py [hunk 2]
```diff
@@ -77,24 +76,60 @@ def device_capabilities() -> DeviceCapabilities:
...
+
+    # Extract relevant information
+    model_id = re.search(r"Model Name: (.*)", hw_info).group(1).strip()
+    chip_id = "Unknown Chip"
```
Comment (AlexCheema) [inline exo/topology/device_capabilities.py:90]: Shouldn't we start by trying to parse this from the `system_profiler` response?

File: exo/topology/device_capabilities.py [hunk 3]
```diff
@@ -77,24 +76,60 @@ def device_capabilities() -> DeviceCapabilities:
     else:
         return DeviceCapabilities(model=f"Unknown Device", chip=f"Unknown Chip", memory=psutil.virtual_memory().total // 2**20, flops=DeviceFlops(fp32=0, fp16=0, int8=0))
 
+
+import subprocess
+import re
```
Comment (AlexCheema) [inline exo/topology/device_capabilities.py:81]: these imports at top

File: exo/topology/device_capabilities.py (hunk 1 above)
Comment (stephanj) [reply exo/topology/device_capabilities.py]: Indeed, removed "re" and used json processing instead: 

```json
system_profiler SPHardwareDataType -json
{
  "SPHardwareDataType" : [
    {
      "_name" : "hardware_overview",
      "activation_lock_status" : "activation_lock_disabled",
      "boot_rom_version" : "10151.101.3",
      "chip_type" : "Unknown",
      "machine_model" : "MacBookPro18,2",
      "machine_name" : "MacBook Pro",
      "model_number" : "Z14X000PRFN/A",
      "number_processors" : "proc 10:8:2",
      "os_loader_version" : "10151.101.3",
      "physical_memory" : "64 GB",
      "platform_UUID" : "AE593C1A-1051-5AE0-AA12-69CE572FB403",
      "provisioning_UDID" : "AE593C1A-1051-5AE0-AA12-69CE572FB403",
      "serial_number" : "WR295HC54D"
    }
  ]
}
```

File: exo/topology/device_capabilities.py (hunk 2 above)
Comment (stephanj) [reply exo/topology/device_capabilities.py:91]: Done, see latest commit

File: exo/topology/device_capabilities.py (hunk 3 above)
Comment (stephanj) [reply exo/topology/device_capabilities.py]: Replaced regex module "re" with "json"

**Comment by stephanj** [comment]
Looks good on my MacBook Pro :) 

image

**Comment by AlexCheema** [comment]
Looks like you might've accidentally committed your `.idea` :)

File: exo/topology/device_capabilities.py [hunk 4]
```diff
@@ -77,24 +80,59 @@ def device_capabilities() -> DeviceCapabilities:
...
+    model_id = hardware_data.get('machine_model', 'Unknown Model')
+    chip_id = "Unknown Chip"
+    cores = hardware_data.get('number_processors', 0)
+    memory = hardware_data.get('physical_memory', '0 GB').split()[0]
+    memory = int(memory) * 1024  # Convert GB to MB
+
+    # Try to identify the chip using sysctl
+    try:
+        sysctl_output = subprocess.check_output(['sysctl', '-n', 'machdep.cpu.brand_string']).decode('utf-8').strip()
+        if "Apple M1" in sysctl_output:
```
Comment (AlexCheema) [inline exo/topology/device_capabilities.py:99]: Can we generalise to "Apple M{N}" where N can be 1,2,3,....

File: exo/topology/device_capabilities.py (hunk 2 above)
Comment (AlexCheema) [reply exo/topology/device_capabilities.py:91]: Really? It still says `chip_id = "Unknown Chip"`. I don't see where it sets the `chip_id` based on `system_profiler` response.

**Comment by AlexCheema** [comment]
Please also remove `output.json` from VC

**Comment by stephanj** [comment]
Update mac_device_capabilities tests for JSON parsing and mocking

- Refactor tests to use JSON-based system_profiler output
- Add mocked test for MacBook Pro with M3 Max chip
- Improve robustness of chip detection in mac_device_capabilities function
- Update assertions to match new JSON parsing logic

image

**Comment by stephanj** [comment]
Needs some more work because now I'm getting again 0 TFLOPS... hold on

**Comment by stephanj** [comment]
Should be good now! 
image
image

**Comment by AlexCheema** [comment]
Lot of configuration files were committed here.
Closing for now as these values were updated since this PR.

**Pull request closed without merging**